### PR TITLE
Implement global function pointer cache in cnex

### PR DESCRIPTION
**Pull request by gitlarryf**

### CODE REVIEW

File: exec/cnex/cnex.c [hunk 1]
```diff
@@ -224,6 +224,11 @@ int main(int argc, char* argv[])
                         "Max Framesets          : %d\n"
                         "Max Allocations        : %zu\n"
                         "Remaining Objects      : %zu\n"
+                        "Total Predefined Calls : %zu\n"
+                        "Predef Linerar lookups : %zu\n"
```
Comment (ghewgill) [inline exec/cnex/cnex.c:228]: Spelling "Linerar"

File: exec/cnex/cnex.c [hunk 2]
```diff
@@ -1371,10 +1388,19 @@ void exec_CALLP(TExecutor *self)
     _IDEC_glbflags = 0;
     unsigned int start_ip = self->ip;
     self->ip++;
+     g_executor->diagnostics.total_predef_calls++;
```
Comment (ghewgill) [inline exec/cnex/cnex.c:1391]: This line does not appear to be indented correctly

File: exec/cnex/global.c [hunk 3]
```diff
@@ -481,19 +481,19 @@ Cell *global_getVariable(const char *pszVar)
...
+void (*global_lookupFunction(const char *pszFunc, uint32_t *index))(struct tagTExecutor *exec)
 {
     uint32_t i;
 
     i = 0;
     while (gfuncDispatch[i].name) {
         if (strcmp(pszFunc, gfuncDispatch[i].name) == 0) {
-            (*gfuncDispatch[i].func)(exec);
-            return;
+            *index = i;
```
Comment (ghewgill) [inline exec/cnex/global.c:491]: I'm unclear here on why you need to return both the function pointer *and* its index. Wouldn't it be enough to just return one of them?

File: exec/cnex/module.c [hunk 4]
```diff
@@ -114,6 +123,15 @@ TModule *module_loadModule(TExecutor *exec, const char *name, unsigned int modul
     path_readModule(r);
     bytecode_loadBytecode(r->bytecode, r->code, r->codelen);
 
+    // Allocate the global function pointer cache for this module
```
Comment (ghewgill) [inline exec/cnex/module.c:126]: This block of code looks substantially similar to the above block for the main module. Is there a way to consolidate this code into one function?

**Review by ghewgill — CHANGES_REQUESTED** [review]

File: exec/cnex/global.c (hunk 3 above)
Comment (gitlarryf) [reply exec/cnex/global.c:491]: Returning the index is just a diagnostic thing.  It tells me how many linear lookups we had to do to execute the code.  It's not _entirely_ necessary, but I'd like to keep it there, at least for a while.  Probably pull it out at some point.

File: exec/cnex/cnex.c (hunk 1 above)
Comment (gitlarryf) [reply exec/cnex/cnex.c:228]: Fixed.

File: exec/cnex/cnex.c (hunk 2 above)
Comment (gitlarryf) [reply exec/cnex/cnex.c:1391]: Fixed.

File: exec/cnex/module.c (hunk 4 above)
Comment (gitlarryf) [reply exec/cnex/module.c:126]: Yea.  Removed it from module entirely, and moved it to cnex.c.

File: exec/cnex/global.c (hunk 3 above)
Comment (gitlarryf) [reply exec/cnex/global.c:491]: My only other option is to continue to pass a pointer to the `TExecutor` into this function, so I don't have to return the index.  Is that the better option?

Comment (gitlarryf) [reply exec/cnex/global.c]: Fixed.

**Pull request merged**